pod/perlapio.pod

```diff
@@ -69,52 +69,9 @@ perlapio - perl's IO abstraction interface.
   void    PerlIO_debug(const char *fmt,...);
 
 =for apidoc_section Input/Output
-
-=for apidoc Amh|int     |PerlIO_apply_layers|PerlIO *f|const char *mode|const char *layers
-=for apidoc Amh|int     |PerlIO_binmode|PerlIO *f|int ptype|int imode|const char *layers
-=for apidoc ATmh|int    |PerlIO_canset_cnt|PerlIO *f
-=for apidoc Amh|void    |PerlIO_debug|const char *fmt|...
-=for apidoc ATmh|FILE  *|PerlIO_exportFILE|PerlIO *f|const char *mode
-=for apidoc ATmh|int    |PerlIO_fast_gets|PerlIO *f
-=for apidoc ATmh|PerlIO*|PerlIO_fdopen|int fd|const char *mode
-=for apidoc ATmh|FILE  *|PerlIO_findFILE|PerlIO *f
-=for apidoc ATmh|int    |PerlIO_getc|PerlIO *d
-=for apidoc ATmh|int    |PerlIO_getpos|PerlIO *f|SV *save
-=for apidoc ATmh|int    |PerlIO_has_base|PerlIO *f
-=for apidoc ATmh|int    |PerlIO_has_cntptr|PerlIO *f
-=for apidoc ATmh|PerlIO*|PerlIO_importFILE|FILE *stdio|const char *mode
-=for apidoc ATmh|PerlIO*|PerlIO_open|const char *path|const char *mode
-=for apidoc Amh|int     |PerlIO_printf|PerlIO *f|const char *fmt|...
-=for apidoc ATmh|int    |PerlIO_putc|PerlIO *f|int ch
-=for apidoc ATmh|int    |PerlIO_puts|PerlIO *f|const char *string
-=for apidoc ATmh|void   |PerlIO_releaseFILE|PerlIO *f|FILE *stdio
-=for apidoc Amh|PerlIO *|PerlIO_reopen|const char *path|const char *mode|PerlIO *old
-=for apidoc ATmh|void   |PerlIO_rewind|PerlIO *f
-=for apidoc ATmh|int    |PerlIO_setpos|PerlIO *f|SV *saved
-=for apidoc Amh|int     |PerlIO_stdoutf|const char *fmt|...
-=for apidoc ATmh|int    |PerlIO_ungetc|PerlIO *f|int ch
-=for apidoc ATmh|int    |PerlIO_vprintf|PerlIO *f|const char *fmt|va_list args
-
-=for apidoc PerlIO_stdin
-=for apidoc PerlIO_stdout
-=for apidoc PerlIO_stderr
-=for apidoc PerlIO_close
-=for apidoc PerlIO_write
-=for apidoc PerlIO_flush
-=for apidoc PerlIO_eof
-=for apidoc PerlIO_error
-=for apidoc PerlIO_clearerr
-=for apidoc PerlIO_read
-=for apidoc PerlIO_fileno
-=for apidoc PerlIO_setlinebuf
-=for apidoc PerlIO_tell
-=for apidoc PerlIO_seek
-=for apidoc PerlIO_get_cnt
-=for apidoc PerlIO_get_ptr
-=for apidoc PerlIO_set_ptrcnt
-=for apidoc PerlIO_set_cnt
-=for apidoc PerlIO_get_base
-=for apidoc PerlIO_get_bufsiz
+=for apidoc Ay|        |IO
+An internal Input/Output (C<I/O>) object to process input from, or
+or output to, a file or device.
```

Comment:
I'm not sure that's a good characterization, they're about as internal as any other SV variant other than INVLIST (which I believe is meant to be internal/re.pm only)

As with other SV types I believe as long as you use the right macros (the Io*() macros for IO) or functions you're following the API.

Reply:
The problem is I can't find a public macro or fcn that takes such a parameter.  sv_2io returns one, but then what can you do with it?

Reply:
Also GvIO and GvIOp return this, but again what can be done with it?

Reply:
You can fetch the input and output PerlIO * handles with IoIFP(io) and IoOFP(io), which are used in the standard typemap in `lib/ExtUtils/typemap` and in many CPAN modules.  A few CPAN modules use IoDIRP(io) to get the directory handle.

IO.xs uses IoFLAGS(io) to set taint mode and a small number of other CPAN modules also use it.  A few cpan modules use IoTYPE(io).

There's a few more used for storing format related information which only appear to be used in core.

Unfortunately I don't think any of them are documented.

Reply:
How does this revision sound?

Reply:
It feels a bit vague.  How about:

```
An SV type representing I/O.  May contain input and output L<PerlIO|perlapio> objects or
a C<DIR *> from opendir().

It also contains all of the other state associated with a handle including the flush flag, current line
number and L<format|perlform> related information.
```

though some of the phrasing feels wrong.